src/dotnet/commands/dotnet-add-p2p/Program.cs

```diff
@@ -0,0 +1,232 @@
+
+                if (referenceOption.Values.Count == 0)
+                {
+                    throw new GracefulException("You must specify at least one reference to add.");
+                }
+
+                AddProjectToProjectReference(project, frameworkOption.Value(), referenceOption.Values);
+
+                project.Save();
+
+                return 0;
+            });
+
+            try
+            {
+                return app.Execute(args);
+            }
+            catch (GracefulException e)
+            {
+                Reporter.Error.WriteLine(e.Message);
+                return 1;
+            }
+        }
+
+        // There is ProjectRootElement.TryOpen but it does not work as expected
+        // I.e. it returns null for some valid projects
+        public static ProjectRootElement TryOpenProject(string filename)
+        {
+            try
+            {
+                return ProjectRootElement.Open(filename);
+            }
+            catch (Microsoft.Build.Exceptions.InvalidProjectFileException)
+            {
+                return null;
+            }
+        }
+
+        public static ProjectRootElement GetProjectFromFileOrThrow(string filename)
+        {
+            if (!File.Exists(filename))
+            {
+                throw new GracefulException($"Provided project `{filename}` does not exist.");
+            }
+
+            var project = TryOpenProject(filename);
+            if (project == null)
+            {
+                throw new GracefulException($"Invalid MSBuild project `{filename}`.");
+            }
+
+            return project;
+        }
+
+        public static ProjectRootElement GetProjectFromCurrentDirectoryOrThrow()
+        {
+            DirectoryInfo currDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+            FileInfo[] files = currDir.GetFiles("*proj");
+            if (files.Length == 0)
+            {
+                throw new GracefulException("Could not find any MSBuild project in the current directory.");
+            }
+
+            if (files.Length > 1)
+            {
+                throw new GracefulException("Found more than one MSBuild project in the current directory. Please specify which one to use.");
+            }
+
+            FileInfo projectFile = files.First();
+
+            if (!projectFile.Exists)
+            {
+                throw new GracefulException("Could not find any project in the current directory.");
+            }
+
+            var ret = TryOpenProject(projectFile.FullName);
+            if (ret == null)
+            {
+                throw new GracefulException($"Found an MSBuild project `{projectFile.FullName}` in the current directory but it is invalid.");
+            }
+
+            return ret;
+        }
+
+        public static Func<T, bool> AndPred<T>(params Func<T, bool>[] preds)
+        {
+            return (el) => preds.All((pred) => pred == null || pred(el));
```

Comment:
I.e. if you have a function which generates a predicate from an argument and the argument is null which means that we don't want to do the check at all I expect the predicate to be null or always return true since there is nothing to check. it seems more convenient to just treat null as () => true in most of the cases. I was thinking about this function as a general purpose function and not just something used in my context where the null check will never happen. (i.e. if you want to use that in sorting algorithm and want to pass LESSTHAN function you can do AndPred(gencond(field1), gencond(field2)) without doing something like: AndPred(gencond(field1) ?? () => true, gencond(field2) ?? () => true) which would be both slower and less convenient